src/swig/CMakeLists.txt

```diff
@@ -58,9 +66,54 @@ if (GUILE_FOUND)
 
   file (RELATIVE_PATH _REL_GUILE_SITE_PATH ${GUILE_ROOT_DIR} ${GUILE_SITE_DIR})
   set (GUILE_SITE_PATH ${_REL_GUILE_SITE_PATH})
-  install (FILES ${CMAKE_CURRENT_BINARY_DIR}/nlopt.scm DESTINATION ${GUILE_SITE_PATH})
+  install (FILES ${CMAKE_CURRENT_BINARY_DIR}/guile/nlopt.scm DESTINATION ${GUILE_SITE_PATH})
 
   file (RELATIVE_PATH _REL_GUILE_EXTENSION_PATH ${GUILE_ROOT_DIR} ${GUILE_EXTENSION_DIR})
   set (GUILE_EXTENSION_PATH ${_REL_GUILE_EXTENSION_PATH})
   install (TARGETS nlopt_guile LIBRARY DESTINATION ${GUILE_EXTENSION_PATH})
 endif ()
+
+
+if (JNI_FOUND AND Java_FOUND AND SWIG_FOUND)
+
+  include (UseJava)
+
+  set (SWIG_MODULE_nlopt_java_EXTRA_DEPS nlopt-java.i generate-cpp)
+  set (CMAKE_SWIG_FLAGS -package nlopt)
+
+  # swig_add_module is deprecated
+  # OUTPUT_DIR is ${CMAKE_CURRENT_BINARY_DIR}/java/ + the -package above (with
+  # any '.' replaced by '/'). It must also match the GLOB in glob_java.cmake.
+  swig_add_library (nlopt_java LANGUAGE java SOURCES nlopt.i
+                               OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}/java/nlopt
+                               OUTFILE_DIR ${CMAKE_CURRENT_BINARY_DIR})
+
+  set (CMAKE_SWIG_FLAGS)
+
+  swig_link_libraries (nlopt_java ${nlopt_lib})
+  target_link_libraries (nlopt_java JNI::JNI)
+
+  set_target_properties (nlopt_java PROPERTIES OUTPUT_NAME nloptjni)
+  if (CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+    set_target_properties (nlopt_java PROPERTIES
+                           COMPILE_OPTIONS "-fno-strict-aliasing")
+  endif ()
+
+  install (TARGETS nlopt_java LIBRARY DESTINATION ${NLOPT_INSTALL_LIBDIR})
+
+  # unfortunately, SWIG will not tell us which .java files it generated, so we
+  # have to find out ourselves - this is the only portable way to do so
+  # (The nlopt*.i dependencies are there to force updating the list of sources
+  # on any changes to the SWIG interface code, they are not direct inputs.)
+  add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/java_sources.txt
+                     COMMAND ${CMAKE_COMMAND}
+                             -DBINARY_DIR=${CMAKE_CURRENT_BINARY_DIR}
+                             -P ${CMAKE_CURRENT_SOURCE_DIR}/glob_java.cmake
+                     DEPENDS "${swig_generated_file_fullname}"
+                             nlopt.i nlopt-exceptions.i nlopt-java.i
+                             nlopt_java_swig_compilation glob_java.cmake)
+
+  add_jar (nlopt_jar SOURCES @${CMAKE_CURRENT_BINARY_DIR}/java_sources.txt
+                     OUTPUT_NAME nlopt)
+  install_jar (nlopt_jar ${CMAKE_INSTALL_DATADIR}/java)
```

Comment:
the .java files must be installed too, maybe with the "install(CODE" signature as the list must be read from java_sources.txt via another cmake script